compiler/shared_ast/typing.ml

```diff
@@ -18,177 +18,164 @@
+  let add_var v tau t = { t with vars = Var.Map.add v tau t.vars }
 
-and naked_typ =
-  | TLit of A.typ_lit
-  | TArrow of unionfind_typ list * unionfind_typ
-  | TTuple of unionfind_typ list
-  | TStruct of A.StructName.t
-  | TEnum of A.EnumName.t
-  | TOption of unionfind_typ
-  | TArray of unionfind_typ
-  | TDefault of unionfind_typ
-  | TAny of Any.t
-  | TClosureEnv
+  let add_scope_var v typ t =
+    { t with scope_vars = ScopeVar.Map.add v typ t.scope_vars }
 
-let rec typ_to_ast ~(flags : flags) (ty : unionfind_typ) : A.typ =
-  let typ_to_ast = typ_to_ast ~flags in
-  let ty, pos = UnionFind.get (UnionFind.find ty) in
-  match ty with
-  | TLit l -> A.TLit l, pos
-  | TTuple ts -> A.TTuple (List.map typ_to_ast ts), pos
-  | TStruct s -> A.TStruct s, pos
-  | TEnum e -> A.TEnum e, pos
-  | TOption t -> A.TOption (typ_to_ast t), pos
-  | TArrow (t1, t2) -> A.TArrow (List.map typ_to_ast t1, typ_to_ast t2), pos
-  | TArray t1 -> A.TArray (typ_to_ast t1), pos
-  | TDefault t1 -> A.TDefault (typ_to_ast t1), pos
-  | TAny _ ->
-    if flags.fail_on_any then
-      (* No polymorphism in Catala: type inference should return full types
-         without wildcards, and this function is used to recover the types after
-         typing. *)
-      Message.error ~pos
-        "Internal error: typing at this point could not be resolved"
-    else A.TAny, pos
-  | TClosureEnv -> TClosureEnv, pos
+  let add_scope scope_name ~vars ~in_vars t =
+    {
+      t with
+      scopes = ScopeName.Map.add scope_name vars t.scopes;
+      scopes_input = ScopeName.Map.add scope_name in_vars t.scopes_input;
+    }
 
-let rec ast_to_typ (ty : A.typ) : unionfind_typ =
-  let ty' =
-    match Mark.remove ty with
-    | A.TLit l -> TLit l
-    | A.TArrow (t1, t2) -> TArrow (List.map ast_to_typ t1, ast_to_typ t2)
-    | A.TTuple ts -> TTuple (List.map ast_to_typ ts)
-    | A.TStruct s -> TStruct s
-    | A.TEnum e -> TEnum e
-    | A.TOption t -> TOption (ast_to_typ t)
-    | A.TArray t -> TArray (ast_to_typ t)
-    | A.TDefault t -> TDefault (ast_to_typ t)
-    | A.TAny -> TAny (Any.fresh ())
-    | A.TClosureEnv -> TClosureEnv
-  in
-  UnionFind.make (Mark.copy ty ty')
+  let add_toplevel_var v typ t =
+    { t with toplevel_vars = TopdefName.Map.add v typ t.toplevel_vars }
 
-(** {1 Types and unification} *)
+  let open_scope scope_name t =
+    let scope_vars =
+      ScopeVar.Map.disjoint_union t.scope_vars
+        (ScopeName.Map.find scope_name t.scopes)
+    in
+    { t with scope_vars }
+
+  let dump ppf env =
+    let pp_sep = Format.pp_print_space in
+    Format.pp_open_vbox ppf 0;
+    (* Format.fprintf ppf "structs: @[<hov>%a@]@,"
+     *   (StructName.Map.format_keys ~pp_sep) env.structs;
+     * Format.fprintf ppf "enums: @[<hov>%a@]@,"
+     *   (EnumName.Map.format_keys ~pp_sep) env.enums;
+     * Format.fprintf ppf "vars: @[<hov>%a@]@,"
+     *   (Var.Map.format_keys ~pp_sep) env.vars; *)
```

Comment:
Useful to leave this here for debug?

Reply:
Yes `Env.dump` is not used anyway, but sometimes useful for debugging... which means there is actually no reason to keep it commented 😅